_posts/2019-06-04-loading-large-image-data-with-dask-array.md

```diff
@@ -0,0 +1,150 @@
+steps, it may be very difficult if not impossible to reuse an existing
+pipeline. In other words, there is friction between the loading and analysis
+steps, which frustrates efforts to make things reusable.
+
+Having a modular and general way to load data makes it easy to present data
+stored differently in a standard way. Further having a standard way to present
+data to analytical pipelines allows that part of the pipeline to focus on what
+it does best, analysis! In general, this should decouple these to components in
+a way that improves the experience of users involved in all parts of the
+pipeline.
+
+
+Loading image data with Dask
+----------------------------
+
+A pretty typical case to see is a large directory of image files. In particular
+TIFF files are quite common, though other formats could show up as well. It is
+common to see a naming structure that mirrors that shown below. The `<i>`
+represent different numbers.
+
+```
+mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif
+```
+
+To load a single one of these a user might right some code like this:
+
+```python
+from skimage.io import imread
+a = imread("mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif")
+```
+
+While this works fine for reading a single image into memory, this will quickly
+become prohibitive once we try to load more data than we have memory. So we
+might replace this with something like this.
+
+```python
+from dask.array.image import imread
+a = imread("mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif")
+```
+
+This offloads the actual image reading to scikit-image under the hood (much
+like what happened explicitly before). Though this doesn't keep the data in
+memory. Instead we briefly load the image once to figure out the metadata
+(shape and type of the array loaded) and then drop it from memory. Once we
+actually need this data for computation of some kind, it will be loaded back
+in.
+
+Though it's possible we run into data that we want to load that scikit-image
+doesn't know to handle. In this case we have a few options. Though one pretty
+easy one is we could supply our own [`imread` function for Dask to use](
+https://docs.dask.org/en/latest/array-api.html#dask.array.image.imread ). This
+way we could manage the subtleties of loading our own data image data format.
+It might involve accessing some database to pull some image data out or could
+involve interacting with cloud storage to pull some data down.
+
+Thus far we have only showed how one might load one chunk. To load multiple
+chunks we would need to iterate over our data some how and combine the chunks
+together. Suppose we have one dimension to combine chunks along, we might write
+some code that looks like that below. We would load all of the chunks of data
+we have join them together with Dask Array's `block` to form a larger array. We
+can easily extend this to more dimensions.
+
+```python
+data = []
+for fn in files:
+    data.append(dask.array.imread(fn))
+data = dask.array.block(data)
```

Comment:
I think that `dask.array.block` could use some explanation here.  We might also want to mention `dask.array.concatenate` and `dask.array.stack`.  This might also be a good opportunity to use the `_repr_html_` work, which, fortunately for us, generates HTML that we can plop into this post.

Reply:
(this is probably my most major comment)

One way to structure this post might be as follows:

1.  We got a bunch of interesting looking data from a leading edge microscopist.  Look at this cool image of the data as rendered nicely in 3d (cool image from IndeX)
2.  But!  Before we can analyze this data, we have to get it out of a bunch of TIFF files.  Gross! (truncated output of calling `ls`)
3.  This is actually pretty typical (now that we've hooked the reader, we can add some exposition about imaging data and tiff files and why this is the state of things.  They got a little endorphin boost from the image so we can throw some text at them now)
4.  So lets load this data with NumPy for analysis.  We would normally use scikit-image
5.  But our data is too big!  (maybe some `du -hs` call here on a stack of real image data, like Gokul's)
6.  So we can use Dask arrays to wrap around many TIFF files
7.  Example with dask delayed and da.from_delayed
8.  Now we have this list of lazy single-chunked arrays.  How do we want to combine those together?  Look at da.concatenate and da.stack, perhaps using the `_repr_html_` output (which I think could be very neat!)
9.  But in our case it's actually way more complex than this, because we don't have just one axis along which we want to concatenate/stack things.  This brings us to `dask.array.block`, where we place our delayed images into an array.
10.  Our actual code here looked like this (copy from your existing notebook)
11.  This is kind of gross, but the grossness here reflects the grossness of the layout.  What's nice is that once we do this, we have a nice clean dask array to work from
12.  Now that we've done this, lets write out to an analysis ready format like Zarr.  You can read more about Zarr here ...
13.  We'll pick up from here next time!

Reply:
Any thoughts on this layout @jakirkham ?

Reply:
Thanks for writing this up.

Sorry for being slow. Yeah I have been thinking about this and am still thinking about it.

It's an interesting approach and I can certainly see the appeal. From a high level it seems generally reasonable. Some of the details we may still need to be sorted out.

I'm still trying to figure out how we can leverage as much of the existing text as possible here while moving towards this structure. So have been working through some of the smaller comments that move in this direction.

Will try to restructure things a bit more based on this feedback and then we can see where things stand.

Reply:
Also on the `concatenate`, `stack`, `block` progression, I've added some text and code along these lines. Could use a look when you have a chance.